### PR TITLE
Set Platform.executable and executableArguments for Linux sky_shell

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

**Comment by jason-simmons** [comment]
@abarth

File: shell/platform/linux/dart_platform.cc [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+#define NDEBUG
+
+#include "flutter/shell/platform/linux/dart_platform.h"
+
+#include "dart/runtime/bin/platform.h"
+
+void InitDartPlatform(int argc, char* argv[]) {
+  dart::bin::Platform::SetExecutableName(argv[0]);
+  dart::bin::Platform::SetExecutableArguments(argc - 1, argv);
+}
```
Comment (abarth) [inline shell/platform/linux/dart_platform.cc]: This code doesn't seem to have anything to do with linux.  Maybe move it into common?

File: shell/platform/linux/dart_platform.cc [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#define NDEBUG
```
Comment (abarth) [inline shell/platform/linux/dart_platform.cc]: ?

File: shell/platform/linux/main_linux.cc [hunk 3]
```diff
@@ -157,7 +158,9 @@ int RunInteractive() {
 
 }  // namespace
 
-int main(int argc, const char* argv[]) {
+int main(int argc, char* argv[]) {
+  InitDartPlatform(argc, argv);
```
Comment (abarth) [inline shell/platform/linux/main_linux.cc]: Should we do this on mac too?

File: shell/platform/linux/dart_platform.cc (hunk 1 above)
Comment (jason-simmons) [inline shell/platform/linux/dart_platform.cc]: This requires a platform that provides arguments via argc/argv (i.e. not Android).  Will move this into common for sharing with the MacOS platform

File: shell/platform/linux/dart_platform.cc (hunk 2 above)
Comment (jason-simmons) [inline shell/platform/linux/dart_platform.cc]: This is a workaround for including some semi-internal Dart headers

Comment (jason-simmons) [inline shell/platform/linux/dart_platform.cc]: Moved this into a BUILD file

**Comment by jason-simmons** [comment]
Moved this to shell/platform/common - PTAL

File: shell/platform/common/BUILD.gn [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+source_set("common") {
```
Comment (abarth) [inline shell/platform/common/BUILD.gn]: Sorry, I meant the existing `shell/common` directory.

File: shell/platform/common/BUILD.gn [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  sources = [
+    "dart_platform_init.cc",
+  ]
+
+  defines = []
+  if (is_debug) {
+    defines += ["DEBUG"]
+  } else {
+    defines += ["NDEBUG"]
+  }
```
Comment (abarth) [inline shell/platform/common/BUILD.gn]: This shouldn't be needed.  NDEBUG should already be set properly.  If DEBUG is needed by Dart VM and isn't being set, then we should be able to teach the gn targets inside the dart repo to push that define onto us.

**Comment by jason-simmons** [comment]
Changed this to use the new dart::bin APIs.

PTAL

**Comment by chinmaygarde** [comment]
> Moved this to shell/platform/common - PTAL

I don't see these changes reflected. Did you forget to push the amends maybe?

**Comment by chinmaygarde** [comment]
How about we amend the Shell::InitStandalone API to take an additional vector of command line flags. That way, we can initialize this in `flutter/shell/common`. AFAIK, Android is the only platform where this does not make sense. Flags can be empty there.

**Comment by jason-simmons** [comment]
The shell/platform/common stuff has been superseded.  The latest version just calls two newly added Dart embedder APIs in the Linux platform main().

The Dart APIs basically expect to receive the argv passed into main (at the very least, they need a char** whose lifetime exceeds that of the Dart runtime).  We could have each platform pass argv into Shell::InitStandalone, but imho that will be messier than doing this in main for each applicable platform.

The primary goal of this is running Dart language test suites on Linux sky_shell.  I'm not sure whether Platform.executable/executableArguments is of any use on the mobile platforms.

**Comment by Hixie** [comment]
The framework uses these APIs to provide defaults for the about box, IIRC.

**Pull request merged**